File: Python3/0525. Contiguous Array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        sum_dict = {0:-1}
        temp = 0
        res = 0
        for i in range(len(nums)):
            temp += 1 if nums[i] else -1

            if temp not in sum_dict:
                sum_dict[temp] = i
            else:
                res = max(res,i - sum_dict[temp])
        return res

File: C++/0525. Contiguous Array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int c = 0;
        int res = 0;
        map<int, int> digit;
        digit[0] = -1;
        for(int i = 0; i < nums.size(); i++) {
            if(nums[i] == 0) {
                c--;
            } else {
                c++;
            }
            if(digit.count(c)) {
                res = max(res, i - digit[c]);
            } else {
                digit[c] = i;
            }
        }
        return res;
    }
};
